Make tracks collapsible and improve capture view rendering.

pierricgimmig opened the pull request:
- Add TriangleToggle class
- Clicking on the white triangle on a track will toggle collapse/expand
- Add tab-like track label with rounded corners
- Removed "m_DesiredHeight" as it was introducing non-uniform scale
- Vertical zoom is temporary disabled until properly reimplemented
- "All tracks" track now drawn after GPU tracks

pierricgimmig commented:
![CollapsibleTracks](https://user-images.githubusercontent.com/3687222/81542358-5e1b2180-9329-11ea-865a-0cda78b6b543.png)

ronaldfw commented on OrbitGl/GpuTrack.cpp:
```diff
@@ -20,6 +21,9 @@ GpuTrack::GpuTrack(TimeGraph* time_graph,
   string_manager_ = string_manager;
+
+  // Gpu tracks are collapsed by default.
+  collapse_toggle_.SetActive(false);
```
This feels a bit like the state is the opposite of what I would it expect to be. I read "collapse toggle active" as "UI element is collapsed", but here it's the inverse.

pierricgimmig replied:
I agree, I changed it to an enum which is much more explicit.

ronaldfw commented on OrbitGl/SchedulerTrack.h:
```diff
@@ -12,6 +12,9 @@ class SchedulerTrack : public ThreadTrack {
   float GetHeight() const override;
+
+  protected:
+  float GetYFromDepth(float track_y, uint32_t depth, bool collapsed) override;
```
While this is not a public API, I would prefer not to use booleans in interfaces as much as possible. Using an enum will make it more obvious what is happening at callsites (and avoids the additional /* collapsed */ comment) and also makes it easier to expand in the future (for example, a track could have states COLLAPSED, SUMMARY, and FULL_DETAIL).

ronaldfw commented on OrbitGl/TriangleToggle.h:
```diff
@@ -0,0 +1,46 @@
+class TriangleToggle : public Pickable {
+ public:
+  using StateChangeHandler = std::function<void(bool)>;
+  explicit TriangleToggle(bool initial_state, StateChangeHandler handler,
```
Another case where having a boolean as a function argument (in this case public API), is unclear. Reading this, I don't know what the two possible initial states mean. Is the toggle collapsed or not? Would also suggest to use an enum here.

pierricgimmig replied:
Done.

ronaldfw commented on OrbitGl/Track.cpp:
```diff
@@ -5,73 +5,128 @@
+
+  float increment_radians = 0.5f * M_PI / static_cast<float>(num_sides);
+  for (uint32_t i = 1; i < num_sides; ++i) {
+    float angle = M_PI + static_cast<float>(i) * increment_radians;
```
This line creates a warning (which is treated as error) for me (double to float conversion).

ronaldfw replied:
(And line 33 as well.)

pierricgimmig replied:
Added kPiFloat constant.

ronaldfw commented:
I noticed some bugs when trying this out for myself, probably easier to discuss in person, but in brief: 
- Tracks with dynamically instrumented function events don't collapse for me.
- It's hard to hit the triangle, maybe just make the hitbox larger.
- Some minor layout issues here and there (I'll follow up).

(But overall it's really cool and looks really nice.)

ronaldfw commented:
> * Tracks with dynamically instrumented function events don't collapse for me.

Sorry, this is wrong, I only had a depth of 1, in which case collapsing doesn't do anything. 

(Maybe we should not show the triangle button when collapsing doesn't do anything?)

pierricgimmig commented on OrbitGl/CoreMath.h:
```diff
@@ -3,7 +3,10 @@
 //-----------------------------------
 #pragma once
 
+#define _USE_MATH_DEFINES
```
Yeah let's define our own.  I added kPiFloat.

pierricgimmig commented:
Thanks Ronald for the detailed comments.  I addressed the issues and also added some more roundness on the far right side of the tracks.  Also, the triangle is now a lot easier to toggle, the triangle is greyed out if the tracks can't be collapsed, the GPU tracks are symmetric when collapsed, and various little fixes.

ronaldfw commented:
Very cool, thanks!
